### ISSUE TITLE
ServicePrincipal creating ServicePrincipal - Insufficient privileges to complete the operation

### ISSUE THREAD

**Issue by mohoff** [Azure/azure-cli#12939]
## Describe the bug

As a ServicePrincipal, I want to create another ServicePrincipal by using the command below. The last section contains parts of the debug log.

**Command Name**
`az ad sp create-for-rbac`

**Errors:**
```
Insufficient privileges to complete the operation.
```

## To Reproduce:
The below command is run as SP with all possible roles and directory roles assigned (tried `Global Administrator` too)

- `az ad sp create-for-rbac --skip-assignment --name {} --scopes acrpull --role {} --keyvault {} --create-cert --cert {} --debug`

## Expected Behavior

The above command should create the SP

## Environment Summary
```
macOS-10.15.3-x86_64-i386-64bit
Python 3.8.2

azure-cli 2.2.0 *

```
## Additional Context

The above command in `--debug` mode shows that the actual SP creation succeeds - just the last request, which seems to enable the created SP, fails. This is my interpretation of running `rg "Request body" -A 1` on the debug output, which gives:

```
msrest.http_logger : Request body:
msrest.http_logger : None
--
msrest.http_logger : Request body:
msrest.http_logger : None
--
msrest.http_logger : Request body:
msrest.http_logger : {"availableToOtherTenants": false, "homepage": "...", "keyCredentials": [{"startDate": "2020-04-09T18:39:10.174059Z", "endDate": "2021-04-09T18:39:10.174059Z", "value": "<VALUE>", "keyId": "<KEY-ID>", "usage": "Verify", "type": "AsymmetricX509Cert"}], "displayName": "<NAME>", "identifierUris": ["<URI>"]}
--
msrest.http_logger : Request body:
msrest.http_logger : {"accountEnabled": "True", "appId": "<ID>"}
```

The response to the last request with body `{"accountEnabled": "True", "appId": "<ID>"}` is:

```
{"odata.error":{"code":"Authorization_RequestDenied","message":{"lang":"en","value":"Insufficient privileges to complete the operation."},"requestId":"...","date":"2020-04-09T18:40:10"}}
msrest.exceptions : Insufficient privileges to complete the operation.
Creating service principal failed for appid '<APP-ID>'. Trace followed:
{'Cache-Control': 'no-cache', 'Pragma': 'no-cache', 'Content-Type': 'application/json; odata=minimalmetadata; streaming=true; charset=utf-8', 'Expires': '-1', 'ocp-aad-diagnostics-server-name': '...', 'request-id': '..., 'client-request-id': '...', 'x-ms-dirapi-data-contract-version': '1.6', 'ocp-aad-session-key': '...', 'DataServiceVersion': '3.0;', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains', 'Access-Control-Allow-Origin': '*', 'X-AspNet-Version': '4.0.30319', 'X-Powered-By': 'ASP.NET', 'Duration': '2215079', 'Date': 'Thu, 09 Apr 2020 18:40:10 GMT', 'Content-Length': '219'}
```
The log output ends with
```
cli.azure.cli.core.util : Insufficient privileges to complete the operation.
Insufficient privileges to complete the operation.
```

**Comment by yonzhan** [comment]
add to S169

**Comment by mohoff** [comment]
It turned out that the permission `Directory.Read.All` was missing for the SP.

The support team provided the following steps, which solved the problem:

> For setting API permissions, you would need to access portal.azure.com – Azure Active Directory – App registrations – the application that you are using to make this call – API permissions – Add a permission – Azure
 Active Directory Graph (on the lower part of this list) – Delegated or application permissions, depending on the context in which you are making the call – Directory – Directory.Read.All – Add permissions.
After adding these permissions, you would need to grant admin consent for this tenant to this app by clicking the “Grant admin consent for <tenant name>” in API permissions.

**Comment by jiasli** [comment]
`Global Administrator` is only available for users, not Service Principals. 

The failed request you mentioned is a POST request, so I don't think it is relevant to `Directory.Read.All`. 

```
msrest.http_logger : Request URL: 'https://graph.windows.net/54826b22-38d6-4fb2-bad9-b7b93a3e9c5a/servicePrincipals?api-version=1.6'
msrest.http_logger : Request method: 'POST'
msrest.http_logger : Request body:
msrest.http_logger : {"accountEnabled": "True", "appId": "e3a8ef77-e7cf-422c-b2d8-5f27b1ca0837"}
```

In my test, the only permission a Service Principal need to create another Service Principal is **Azure Active Directory Graph -> Application Permissions -> Application.ReadWrite.OwnedBy** (or the higher **Application.ReadWrite.All**):

![image](https://user-images.githubusercontent.com/4003950/79042306-2be2ab00-7c29-11ea-8d8f-4c64d64aa9ab.png)

After assigning this permission and granting admin consent:

```
> az account show
{
  ...
  "tenantId": "54826b22-38d6-4fb2-bad9-b7b93a3e9c5a",
  "user": {
    "name": "22272fc1-8d0d-4da2-a70b-a44514d0aca4",
    "type": "servicePrincipal"
  }
}

> az ad sp create-for-rbac --skip-assignment
{
  "appId": "f8362dc6-8b5c-474b-9ccc-9c6851b96f48",
  "displayName": "azure-cli-2020-04-11-11-21-16",
  "name": "http://azure-cli-2020-04-11-11-21-16",
  "password": "...",
  "tenant": "54826b22-38d6-4fb2-bad9-b7b93a3e9c5a"
}
```

**Comment by mohoff** [comment]
@jiasli Thanks a lot for your reply, much appreciated. I would like to address the three points you made to understand better the AD and related concepts.

> Global Administrator is only available for users, not Service Principals

Hm, I _can_ assign a SP any role in the Portal: `Active Directory` > `Roles and Administrators` > click any listed role > `Add assignments` > assign Directory Role to SP (works). How can I understand your comment?

> The failed request you mentioned is a POST request, so I don't think it is relevant to Directory.Read.All.

While I'd agree in theory, it turned out that add just this permission solved the issue for me. This could be related to the pre-assigned Directory Roles the SP was already assigned with.

> In my test, the only permission a Service Principal need to create another Service Principal is Azure Active Directory Graph -> Application Permissions -> Application.ReadWrite.OwnedBy 

Nice, works for me too. This should be the better choice.

**Comment by jiasli** [comment]
Hi @mohoff, I got your point. I just found adding Service Principal is recently discussed at https://github.com/MicrosoftDocs/azure-docs/issues/49478. I followed your steps and reproduced the issue.

### Step 1. Add Role Assignments 

**Azure Active Directory > Roles and Administrators > Global administrator > Add assignments > assign Directory Role to SP**

![image](https://user-images.githubusercontent.com/4003950/79179178-9ec27080-7e39-11ea-8e22-ce3832b7ab4b.png)

I got the same error when running 

```
> az ad sp create-for-rbac --skip-assignment --debug
msrest.http_logger : Request URL: 'https://graph.windows.net/54826b22-38d6-4fb2-bad9-b7b93a3e9c5a/servicePrincipals?api-version=1.6'
msrest.http_logger : Request method: 'POST'
msrest.http_logger : Request body:
msrest.http_logger : {"accountEnabled": "True", "appId": "54886d52-2765-4f33-879d-ad170daa8bda"}
msrest.http_logger : Response status: 403
msrest.http_logger : Response content:
msrest.http_logger : {"odata.error":{"code":"Authorization_RequestDenied","message":{"lang":"en","value":"Insufficient privileges to complete the operation."},"requestId":"555f6bc3-89f5-4917-9f3c-7e032d99f98e","date":"2020-04-14T01:48:45"}}
```

### Step 2. Add API Permissions 

**Azure Active Directory > App registrations > select my app > API Permissions > Azure Active Directory Graph -> Application Permissions -> Directory.Read.All**

![image](https://user-images.githubusercontent.com/4003950/79179206-b13caa00-7e39-11ea-9676-153646563cf6.png)

Then `az ad sp create-for-rbac --skip-assignment` starts to work. 

So I think the questions is 

1. Why does **Global Administrator** given to a SP still not have enough permission to create an app, but fails on a `POST` request which needs `Directory.Read.All` permission?
2. What is the difference between [Role Permissions](https://docs.microsoft.com/en-us/azure/active-directory/users-groups-roles/directory-assign-admin-roles) and [API Permissions](#issuecomment-612398494)?

Let me sync with AAD team internally and get back to you.

**Comment by mohoff** [comment]
Thanks @jiasli , good to see you could reproduce.

Also great questions. I'm generally confused with different kinds of permissions for different APIs (Microsoft Graph vs AAD Graph) and what is supported by the `az` CLI tool.

**Comment by mohoff** [comment]
@jiasli Any update on this?

**Comment by jiasli** [comment]
Thanks for checking. We are still communicating with AAD team. Thanks for your patience. But for now, let use it as it is to get unblocked. 😉

**Comment by jiasli** [comment]
@mohoff, as I tested again, creating Service Principal using a **Global administrator** Service Principal now doesn't require `Directory.Read.All` anymore. It looks like the service has been changed recently. Could you try again? (Please note that role membership changes take some time (around 10min) to propagate.)

**Comment by eugeneromero** [comment]
I am currently trying to set up a pipeline where a Service Principal has permissions to create other SPs on demand. It appears that with the update from AAD Graph to MS Graph, there is a lot of confusing information online as to how this should properly be set up. Since testing in the corporate environment is difficult, as I would need to constantly be going back to the Azure Admin to get him to Admin Approve my API permission requests, I decided to test in a personal account I control.

First, I created the "top" SP with `az ad sp create-for-rbac --name devopsagent --role owner`. From there, I create a clean environment, install `az cli` and login:

`az login --service-principal -u "devopsagent_appid" -p "devopsagent_pass" --tenant "ad_tenant"`

Now, I try to create a secondary SP:

`az ad sp create-for-rbac --skip-assignment --name limited-sp`

This, as expected, fails:
`ValidationError: Insufficient privileges to complete the operation.`

At this point, I started trying to find the minimum set of permissions that would get this working. This is where my confusion is (and why I am adding to this issue):

The Azure portal recommends using Microsoft Graph API permissions, instead of Azure Active Directory Graph, which is now on life support. So I try adding these two MS Graph permissions in the portal:

```
Microsoft Graph/Application permissions/Application/Application.ReadWrite.OwnedBy
Microsoft Graph/Application permissions/Directory/Directory.Read.All
```

This does **not** work. I get either:
```
ValidationError: Insufficient privileges to complete the operation.
```
or (not entirely sure why the error changes, maybe because of back-and-forth with permissions)
```
ValidationError: Directory permission is needed for the current user to register the application. For how to configure, please refer 'https://docs.microsoft.com/azure/azure-resource-manager/resource-group-create-service-principal-portal'. Original error: Insufficient privileges to complete the operation.
```
I tried changing the `Directory.Read.All` to `Directory.ReadWriteAll`, same result. Additionally, I tried adding `Directory.ReadWriteAll` from the AAD Graph API, same result. The only way I can get it to work, is adding these two permissions:
```
Azure Active Directory Graph/Application permissions/Application/Application.ReadWrite.OwnedBy
Azure Active Directory Graph/Application permissions/Directory/Directory.Read.All
```

This makes the request work. Most interestingly, **removing the MS Graph permissions** and only leaving the AAD ones makes no difference. So as of today, it does not seem that the `az cli` is using the MS Graph API at all, at least for this particular task. Is this correct?

As an additional note, based on previous comments on this issue, **I did not need to add the top SP to any groups** (global admin or others). As mentioned above, even adding to the Global Admins group, I still got an error.

My `az --version` is:
```
# az --version
azure-cli                         2.12.1
core                              2.12.1
telemetry                          1.0.6

Python location '/opt/az/bin/python3'
Extensions directory '/root/.azure/cliextensions'

Python (Linux) 3.6.10 (default, Sep 28 2020, 08:41:31) 
[GCC 9.3.0]
Your CLI is up-to-date.
```
I guess my main question is, will the MS Graph API permissions eventually replace the AAD ones? This is my understanding. So, in preparation and to bother the Azure Admin as little as possible, should I add **both sets** of API permissions? Or is there something I am not getting correctly?

Also, currently using any APIs from the AAD set, pops up this warning in the Azure window, which the Admin *will see* and *will ask about* :laughing: So I guess an answer to my above questions should make for a proper answer for him.

![image](https://user-images.githubusercontent.com/1104426/95321740-c26fec00-089b-11eb-9744-7b1b9a53f2d1.png)

Thank you for your work on this!

**Comment by iTiamo** [comment]
Thanks @eugeneromero... Having to jump through hoops and look at Github issues to fix a problem always makes me feel like I'm doing something unintended.

**Comment by mwindaji-000** [comment]
@iTiamo did you ever get a solution to this problem. I currently having the same issue and am curious how this went.

**Comment by jiasli** [comment]
Hi @eugeneromero, thank you for the detailed explanation. Your statement is correct: **Azure CLI `az ad` command group currently only uses Azure Active Directory Graph, so you need to add Azure Active Directory Graph permissions for `az ad` to work.**

If you are interested in using Microsoft Graph, please add corresponding Microsoft Graph permissions and use `az rest` to make the API calls. Please see https://github.com/Azure/azure-cli/issues/12946 for more detail on the explanation and instructions on using `az rest` with Microsoft Graph.

Azure CLI team is working on migrating `az ad` to use Microsoft Graph, but this is a big task and we can't provide a solid ETA yet. 

Meanwhile, Microsoft Graph team is currently working on their own CLI tool: https://github.com/microsoftgraph/msgraph-cli. This project is still at its early phase. You are very welcome to play with it and share any feedback.

**Comment by chengyu-liu-cs** [comment]
I am having the same issue. Is there complete working solution/example ? @jiasli 

@Michael-Coetzee What else did you do to create a service principal from a service principal beside the two command lines you used ?

> Okaaaay... can confirm, this works
az rest -m POST --uri "https://graph.microsoft.com/v1.0/applications" --body '{ "displayName": "Test-App" }'
az rest -m POST --uri "https://graph.microsoft.com/v1.0/servicePrincipals/" --body '{ "appId": "<appId-from-previous-command>" }'

**Comment by eflorespalma** [comment]
Hi Guys (@jiasli @Michael-Coetzee), I have the same problem that you were talking in this thread, however I did all you suggest and It's working only in windows, but when I tried on docker linux(Ubuntu)  I got the "Insufficient privileges to complete the operation" message. I am using the same credential to create the application. These are the commands I am using for both environments

```
az login --service-principal --username xxxxxx --password xxxxxx --tenant xxxxx
az ad sp create-for-rbac --display-name sa-playground-handler --years 5
```

![image](https://github.com/Azure/azure-cli/assets/2238801/2496d905-374b-4ee6-938d-f4720c2a1fc0)

Working on windows
![image](https://github.com/Azure/azure-cli/assets/2238801/55efccd3-e653-49bd-825f-44d6086bec43)

Error on Docker(Ubuntu)
![image](https://github.com/Azure/azure-cli/assets/2238801/3f8f5127-4eae-4b37-a076-ba3bb76443ef)

Any idea about this behaviour?
